docs/proposals/proxy-extensions.md

```diff
@@ -0,0 +1,294 @@
+---
+title: Reverse Proxy Extensions
+
+authors:
+- "@leoluz"
+
+sponsors:
+- TBD
+
+reviewers:
+- TBD
+
+approvers:
+- TBD
+
+creation-date: 2022-07-23
+
+---
+
+# Reverse-Proxy Extensions support for ArgoCD
+
+Enable UI extensions to use a backend service.
+
+* [Summary](#summary)
+* [Motivation](#motivation)
+* [Goals](#goals)
+* [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [Use cases](#use-cases)
+    * [Security Considerations](#security-considerations)
+    * [Risks and Mitigations](#risks-and-mitigations)
+    * [Upgrade / Downgrade](#upgrade--downgrade)
+* [Drawbacks](#drawbacks)
+* [Open Questions](#open-questions)
+
+---
+
+## Summary
+
+ArgoCD currently supports the creation of [UI extensions][1] allowing
+developers to define the visual content of the "more" tab inside
+a specific resource. Developers are able to access the resource state to
+build the UI. However, currently it isn't possible to use a backend
+service to provide additional functionality to extensions. This proposal
+defines a new reverse proxy feature in ArgoCD, allowing developers to
+create a backend service that can be used in UI extensions.
+
+## Motivation
+
+The initiative to implement the anomaly detection capability in ArgoCD
+highlighted the need to improve the existing UI extensions feature. The
+new capability will required the UI to have access to data that isn't
+available as part of Application's owned resources. It is necessary to
+access an API defined by the extension's development team so the proper
+information can be displayed.
+
+## Goals
+
+The following goals are desired but not necessarily all must be
+implemented in a given ArgoCD release:
+
+#### [G-1] ArgoCD (API Server) must have low performance impact when running extensions
+
+ArgoCD API server is a critical component as it serves all APIs used by
+the CLI as well as the UI. The ArgoCD team has no controll over what is
+going to be executed in extension's backend service. Thus it is important
+that the reverse proxy implementation to cause the lowest possible impact
+in the API server while processing high latency requests.
+
+Possible solutions:
+- reverse proxy implemented as an independent server
+- reverse proxy invoke backend services asynchronously
+
+----
+
+#### [G-2] ArgoCD admins should be able to define rbacs to define which users can invoke specific extensions
+
+ArgoCD Admins must be able to define which extensions are allowed to be
+executed by the logged in user. This should be fine grained by ArgoCD
+project like the current rbac implementation.
+
+----
+
+#### [G-3] ArgoCD deployment should be independent from backend services
```

Comment:
@leoluz Has it been addressed if extensions have the ability to interact with other extensions?

Reply:
Makes sense! Will RBAC permissions will still apply if the extension invokes another extension? ie if a user with a certain permission invokes an action in an extension and that action invokes another action in an extension the user normally wouldn't have access to, will RBAC then deny this action? Or is this too complex for this iteration?

Reply:
> Will RBAC permissions will still apply if the extension invokes another extension?

No. The RBAC will apply for the call between Argo CD in the browser and the API Server. Once the request is forwarded to the extension backend, it is up to that service to implement the authorization the developer see fit.